### PR TITLE
Replica placement improvements

### PR DESCRIPTION
**Pull request by cin**
This PR attempts to make better choices around replica selection when workers aren't spread across multiple zones. The algorithm always tries to pick the replica that is RF # lexicographically greater than the primary (so the first replica is 1 worker over, the. 2nd is 2 workers over, etc). See the examples below.

# 3 Workers with 3 Primaries and RF=1
```
+--------------------------------+       +-------------------------------+      +--------------------------------+
|            worker1             |       |           worker2             |      |            worker3             |
|      +------------------+      |       |      +-----------------+      |      |      +------------------+      |
|      |      redis1      |      |       |      |     redis2      |      |      |      |      redis3      |      |
|      |      primary     |      |       |      |     primary     |      |      |      |      primary     |      |
|      +------------------+      |       |      +-----------------+      |      |      +------------------+      |
|      +------------------+      |       |      +-----------------+      |      |      +------------------+      |
|      |      redis3      |      |       |      |     redis1      |      |      |      |      redis2      |      |
|      |      replica     |      |       |      |     replica     |      |      |      |      replica     |      |
|      +------------------+      |       |      +-----------------+      |      |      +------------------+      |
|                                |       |                               |      |                                |
+--------------------------------+       +-------------------------------+      +--------------------------------+
```

# 3 Workers with 3 Primaries and RF=2
```
+---------------------------------+       +-------------------------------+      +-------------------------------+
|            worker1              |       |           worker2             |      |            worker3            |
|      +-------------------+      |       |      +-----------------+      |      |      +-----------------+      |
|      |      redis1       |      |       |      |     redis2      |      |      |      |      redis3     |      |
|      |      primary      |      |       |      |     primary     |      |      |      |      primary    |      |
|      +-------------------+      |       |      +-----------------+      |      |      +-----------------+      |
|      +-------------------+      |       |      +-----------------+      |      |      +-----------------+      |
|      |      redis3       |      |       |      |     redis1      |      |      |      |      redis2     |      |
|      |      replica      |      |       |      |     replica     |      |      |      |      replica    |      |
|      +-------------------+      |       |      +-----------------+      |      |      +-----------------+      |
|      +-------------------+      |       |      +-----------------+      |      |      +-----------------+      |
|      |      redis2       |      |       |      |     redis3      |      |      |      |      redis1     |      |
|      |      replica      |      |       |      |     replica     |      |      |      |      replica    |      |
|      +-------------------+      |       |      +-----------------+      |      |      +-----------------+      |
|                                 |       |                               |      |                               |
+---------------------------------+       +-------------------------------+      +-------------------------------+
```

### CODE REVIEW

**Review by 4n4nd — COMMENTED** [review]
@cin this looks great! 
One thing I would like to point out is that after it fails to place replicas optimally. When in best effort mode, it should at least try to place replicas on different k8s nodes from primaries. I don't think that is being done here. 
https://github.com/IBM/operator-for-redis-cluster/blob/ef88f333fa8a81870fb0e727e26571b66e831cbf/pkg/controller/clustering/cluster-placement.go#L335

**Comment by cin** [comment]
Correct, currently it just falls back to the old logic if an optimal placement cannot be found. I could make it use the same logic though. All of my current tests have just hit the optimal placement logic. Actually I need to look into when the best effort logic kicks in...

**Comment by 4n4nd** [comment]
> BTW, sorry for the delay on this. We've had some other issues pop up that have been taking my time. I should be able to get back to fixing this next week.

no worries! you have been super helpful 👍 
Please let me know if there's any way I could contribute to this PR if you don't have the cycles.

**Comment by 4n4nd** [comment]
> I could make it use the same logic though.

I don't know if that would work. I don't see k8s nodes being checked in the old logic as well.

**Comment by 4n4nd** [comment]
> it should at least try to place replicas on different k8s nodes from primaries

@cin I tried adding on to your PR to add this functionality. [Here](https://github.com/IBM/operator-for-redis-cluster/compare/51-replica-placement-improvements...4n4nd:operator-for-redis-cluster:pr-51?expand=1) is the code. I tested it and it seems to be working fine for me.

**Comment by cin** [comment]
> > it should at least try to place replicas on different k8s nodes from primaries
> 
> @cin I tried adding on to your PR to add this functionality. [Here](https://github.com/IBM/operator-for-redis-cluster/compare/51-replica-placement-improvements...4n4nd:operator-for-redis-cluster:pr-51?expand=1) is the code. I tested it and it seems to be working fine for me.

@4n4nd Have you seen the BestEffort scheduling kick in? I've only seen the optimal route hit in my testing.

I think I've gotten the replica removal logic to follow the same pattern as the placement. If you get some time, maybe give it a whirl and see if it works for your use case.

**Comment by 4n4nd** [comment]
> @4n4nd Have you seen the BestEffort scheduling kick in? I've only seen the optimal route hit in my testing.

yeah actually. I have a small 4 nodes cluster. I divide the nodes into 2 zones and increase the number of replicas to 3 or more so the operator runs out of new zones and goes into BestEffort mode.

> I think I've gotten the replica removal logic to follow the same pattern as the placement. If you get some time, maybe give it a whirl and see if it works for your use case.

Will give it a go. Thanks!

**Comment by 4n4nd** [comment]
@cin I tried it out and the scheduling seems to be working fine as long as it is using the optimal placement strategy. But during best effort placement, it doesn't really change much.

The following [if statement](https://github.com/IBM/operator-for-redis-cluster/blob/3996836f90b8c6e63c76aed766a635898738db12/pkg/controller/clustering/cluster-placement.go#L321) just skips optimal placement of replicas and goes straight to best effort placement if replication factor is greater than the number of zones. 
```golang
if int(replicationFactor) < len(zones) {
  // skip this zone if it is the same as the primary zone or if this primary already has replicas in this zone
  if primary.Zone == zone || SameZone(zone, primaryToReplicas[primary.ID]) {
	continue
  }
}
``` 

And [best effort placement](https://github.com/IBM/operator-for-redis-cluster/blob/3996836f90b8c6e63c76aed766a635898738db12/pkg/controller/clustering/cluster-placement.go#L338) just picks the first available pod. I think this can be improved by adding a simple loop that checks if any replicas or primaries for a shard are present in the same node (A zone can have multiple k8s nodes).

**Comment by 4n4nd** [comment]
@cin I was thinking instead of the operator managing the pods individually, maybe it would make more sense to use Deployments or Statefulsets. That way we could let the k8s scheduler worry about the pod spread and replica placement. Wdyt?

**Comment by cin** [comment]
@4n4nd Our team talked about doing this a while back, but it will be a lot of work. There are also many unknowns, so trying to estimate the scope of the work would be difficult (we didn't start this project and there are many areas of the code I have not even looked at). So we went with a lighter touch and tried to work with what we have. Statefulsets would probably be a good approach to deal with this but there are still scheduling issues that would be encountered (especially when the number of pods % nodes is not zero and/or nodes are shared). There are also quite a few different ways to use statefulsets here. Either way though, I think we'd still have to deal with picking the right replicas for primaries. I wish you could just set primary-0's replicas to replica1-1, replica2-2, etc but there's no guarantee about where these pods will be scheduled relative to one another. And of course shared nodes throws that off even more.

Another approach would be to holistically consider the cluster when pairing primaries and replicas. The logic right now looks at a primary pod and tries to figure out which set of unassigned replica pods are the best fit for it. Obviously any decisions we make about pairing primaries to replicas early in the cycle impacts which replicas are available later in the cycle. This would also require a fairly large re-write, hence the solution above. This is just trying to deterministically place replicas for primaries. However this breaks down on repeated scaling operations -- as does the current logic to be fair.

We typically don't have to worry about this as we don't share nodes. We just right-size the nodes for how much memory we plan to use (taking into account other system and prometheus pods on the node and at least a 30% overhead depending on the use case). This eliminates nearly all scheduling issues and should allow more efficient redis operations as it's not sharing resources with other redis pods. It's also be more resilient to node failures. This is not the most cost effective solution for non-prod environments but was easiest to implement given the time constraints we were under.

Along these lines, I would also like to schedule all the pods at once (instead of one by one). It would speed up the deployment of a new cluster substantially. If we moved to statefulsets this would actually be easy. I haven't looked into how it could be done in the existing code. It's just been one of those things that's annoyed me.

TLDR: I'm open to moving to statefulsets. It's just the effort required that is problematic. That and you still have to deal with scheduling issues. The 2nd approach mentioned would be easier to implement but would still require touching a lot of logic (and probably updating tests). Either way, I think the logic for picking replicas for primaries needs to change. I don't think the solution in this PR works very well in its current form. There are a few things I can do to make it better, but they're all just work arounds that I think won't work in a deterministic fashion when scaling. I'm open to ideas and suggestions though. :)
